mitodl/ocw-studio
# update ocw import + concourse build time handling of `publish_date` field

#### Pre-Flight checklist


#### What are the relevant tickets?

closes #916

#### What's this PR do?

This PR adjusts the role played by the `publish_date` field in a few ways.

1. Introduce a new field on the `Website` model called `first_published_to_production`. The migration to introduce this field also sets `publish_date` to `None` for all existing `Website` records
2. during the OCW import, instead of setting `Website.publish_date` to the value for `publishdate` on the ocw2hugo data, stick that value into the `first_published_to_production` field instead
3. when doing a Concourse build, update the `Website` record's `publish_date` and whatnot a bit more carefully. In particular, for both `draft` and `live`:
    1. only set `publish_date` to `datetime.now()` when the build finished successfully
    2. if a `Website` record's `.first_publish_to_production` field is `None`, then set that to `datetime.now()` as well

Together this should ensure that we have an unambiguous ability to tell, given the data we store, whether or not a course has been successfully published to production through our Concourse pipeline. More specifically, if a `Website` record has a non-null value for `publish_date` then it has been published through our system. We also, by storing the OCW legacy `publishdate` value preserve information about when those courses have been updated and whatnot.

#### How should this be manually tested?

I think it will be necessary to test out the migration, run a course import, and then test out publishing a course, to make sure that everything works as described above.
